src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1161,6 +1161,77 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
     return parts_with_ranges;
 }
 
+RangesInDataParts MergeTreeDataSelectExecutor::applyLimitForRangesInDataParts(
+    RangesInDataParts && parts_with_ranges,
+    size_t limit)
+{
+    chassert(limit > 0);
+    RangesInDataParts result_parts_with_ranges;
+    size_t remaining_limit = limit;
+
+    for (auto && part_with_ranges : parts_with_ranges)
+    {
+        if (remaining_limit == 0)
+            break;
+
+        size_t part_ranges_total_rows = part_with_ranges.getRowsCount();
+        result_parts_with_ranges.push_back(std::move(part_with_ranges));
+        if (part_ranges_total_rows <= remaining_limit)
+        {
+            remaining_limit -= part_ranges_total_rows;
+            continue;
+        }
+
+        auto & result_part_with_ranges = result_parts_with_ranges.back();
+        const auto & part_index_granularity = result_part_with_ranges.data_part->index_granularity;
+
+        auto ranges_it = result_part_with_ranges.ranges.begin();
+        auto ranges_end_it = result_part_with_ranges.ranges.end();
+        for (; ranges_it != ranges_end_it; ++ranges_it)
+        {
+            const auto & range = *ranges_it;
+            if (remaining_limit == 0)
+                break;
+
+            size_t rows_count_in_range = part_index_granularity.getRowsCountInRange(range);
```

Comment:
How does it work with lightweight deletes? I suspect that this method can return the number of rows together with deleted. Let's check it.

Reply:
Updated.